Add lookahead to reduce/reduce conflicts

quernd opened the pull request:
This adds more diagnostic information to reduce/reduce conflicts based on the observation that different lookahead symbols don't necessarily cause the same conflicts.

For instance, the LR automaton for this grammar:

```
%start S
%%
S: A X | B X | A Y;
A: '0' ; B: '0' ; C: '0' ;
X: '1' ; Y: '2' ;
```

has the following item set for state 1:

```
1:  [B -> '0' ., {'1'}]
    [A -> '0' ., {'1', '2'}]
```

which causes a reduce/reduce conflict if the lookahead is '1', but not if the lookahead '2'.

ltratt commented:
I think the tests need updating as with this commit they won't compile?

```
error[E0277]: can't compare `(TIdx<u32>, PIdx<u32>, PIdx<u32>, StIdx<u32>)` with `(PIdx<u32>, PIdx<u32>, StIdx<{integer}>)`
   --> lrtable/src/lib/statetable.rs:982:9
    |
982 | /         assert_eq!(
983 | |             conflicts.rr_conflicts().next().unwrap(),
984 | |             &(
985 | |                 grm.rule_to_prods(grm.rule_idx("B").unwrap())[0],
...   |
989 | |         );
    | |_________^ no implementation for `(TIdx<u32>, PIdx<u32>, PIdx<u32>, StIdx<u32>) == (PIdx<u32>, PIdx<u32>, StIdx<{integer}>)`
    |
    = help: the trait `PartialEq<(PIdx<u32>, PIdx<u32>, StIdx<{integer}>)>` is not implemented for `(TIdx<u32>, PIdx<u32>, PIdx<u32>, StIdx<u32>)`
    = help: the following other types implement trait `PartialEq<Rhs>`:
              ()
              (A, Z, Y, X, W, V, U, T)
              (B, A, Z, Y, X, W, V, U, T)
              (C, B, A, Z, Y, X, W, V, U, T)
              (D, C, B, A, Z, Y, X, W, V, U, T)
              (E, D, C, B, A, Z, Y, X, W, V, U, T)
              (T,)
              (U, T)
            and 5 others
    = note: required for `&(TIdx<u32>, PIdx<u32>, PIdx<u32>, StIdx<u32>)` to implement `PartialEq<&(PIdx<u32>, PIdx<u32>, StIdx<{integer}>)>`
    = note: this error originates in the macro `assert_eq` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `lrtable` (lib test) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
```

I am also not totally certain off-hand if we have a test which checks the necessary output. If we do, that will obviously need updating. If we don't, it would be good to add a simple test, to make sure we don't regress if/when we merge this.

quernd commented:
Sorry, my bad! I updated the existing tests (as well as a diagnostic line in `nimbleparse`) and added a new test involving the grammar I mentioned above.

ltratt commented:
This looks good to me! @ratmice any thoughts?

ratmice commented:
I haven't had too much digesting of the issue (I'm not as fast as you guys!), I'm curious if there is a way to integrate [token_span](https://docs.rs/cfgrammar/latest/cfgrammar/yacc/grammar/struct.YaccGrammar.html#method.token_span)

for underlining into the nimbleparse diagnostic, I suppose it can always be done in a follow up if it turns out to be beneficial and not too busy?.

ltratt commented:
> I'm curious if there is a way to integrate [token_span](https://docs.rs/cfgrammar/latest/cfgrammar/yacc/grammar/struct.YaccGrammar.html#method.token_span) for underlining into the nimbleparse diagnostic

That would be nice -- and a nice follow-up PR! [Thanks for the suggestion @ratmice!]

@quernd Please squash the commits down and we're ready to merge.

quernd commented:
> > I'm curious if there is a way to integrate [token_span](https://docs.rs/cfgrammar/latest/cfgrammar/yacc/grammar/struct.YaccGrammar.html#method.token_span) for underlining into the nimbleparse diagnostic
> 
> That would be nice -- and a nice follow-up PR! [Thanks for the suggestion @ratmice!]
> 
> @quernd Please squash the commits down and we're ready to merge.

Thanks, squashed down to one commit!

Thanks for the pointer to `token_span` @ratmice. I'll try and follow up.
